src/repr/src/lib.rs

```diff
@@ -51,9 +57,298 @@ pub use scalar::{
+            None => panic!("could not step {self} forward by {amount}"),
+        }
+    }
+
+    /// Retreat a timestamp by the least amount possible such that
+    /// `ts.step_back().unwrap().less_than(ts)` is true. Return `None` if unable,
+    /// which must only happen if the timestamp is `Timestamp::minimum()`.
+    pub fn step_back(&self) -> Option<Self> {
+        self.checked_sub(1)
+    }
+}
+
+// todo comment this out to see if it's used?
+impl mz_proto::RustType<u64> for Timestamp {
+    fn into_proto(&self) -> u64 {
+        self.into()
+    }
+
+    fn from_proto(proto: u64) -> Result<Self, mz_proto::TryFromProtoError> {
+        Ok(proto.into())
+    }
+}
+
+impl From<u64> for Timestamp {
+    fn from(internal: u64) -> Self {
+        Self { internal }
+    }
+}
+
+impl From<Timestamp> for u64 {
+    fn from(ts: Timestamp) -> Self {
+        ts.internal
+    }
+}
+
+impl TryFrom<Timestamp> for i64 {
+    type Error = TryFromIntError;
+
+    fn try_from(value: Timestamp) -> Result<Self, Self::Error> {
+        value.internal.try_into()
+    }
+}
+
+impl From<&Timestamp> for u64 {
+    fn from(ts: &Timestamp) -> Self {
+        ts.internal
+    }
+}
+
+impl From<Timestamp> for Numeric {
+    fn from(ts: Timestamp) -> Self {
+        ts.internal.into()
+    }
+}
+
+impl std::ops::Rem<Timestamp> for Timestamp {
+    type Output = Timestamp;
+
+    fn rem(self, rhs: Timestamp) -> Self::Output {
+        Self {
+            internal: self.internal % rhs.internal,
+        }
+    }
+}
+
+impl timely::order::PartialOrder for Timestamp {
+    fn less_equal(&self, other: &Self) -> bool {
+        self.internal.less_equal(&other.internal)
+    }
+}
+
+impl timely::order::TotalOrder for Timestamp {}
+
+impl timely::progress::Timestamp for Timestamp {
+    type Summary = Timestamp;
+
+    fn minimum() -> Self {
+        Self { internal: u64::MIN }
+    }
+}
+
+impl timely::progress::PathSummary<Timestamp> for Timestamp {
+    #[inline]
+    fn results_in(&self, src: &Timestamp) -> Option<Timestamp> {
+        self.internal
+            .checked_add(src.internal)
+            .map(|internal| Self { internal })
+    }
+    #[inline]
+    fn followed_by(&self, other: &Timestamp) -> Option<Timestamp> {
+        self.internal
+            .checked_add(other.internal)
+            .map(|internal| Self { internal })
+    }
+}
+
+impl timely::progress::timestamp::Refines<()> for Timestamp {
+    fn to_inner(_: ()) -> Timestamp {
+        Default::default()
+    }
+    fn to_outer(self) -> () {
+        ()
+    }
+    fn summarize(_: <Timestamp as timely::progress::timestamp::Timestamp>::Summary) -> () {
+        ()
+    }
+}
+
+impl differential_dataflow::lattice::Lattice for Timestamp {
+    #[inline]
+    fn join(&self, other: &Self) -> Self {
+        ::std::cmp::max(*self, *other)
+    }
+    #[inline]
+    fn meet(&self, other: &Self) -> Self {
+        ::std::cmp::min(*self, *other)
+    }
+}
+
+impl mz_persist_types::Codec64 for Timestamp {
```

Comment:
I believe the actual bytes here are the same as any existing bytes, so those should be ok? Unsure if the `codec_name` matters there or not. @dan?